Introduce LegacyTHDispatcher for dispatching to TH functions.

gchanan opened the pull request:
This isn't hooked up to anything yet, this is just putting the skeleton in place.
The idea here is that the functions generated via Declarations.cwrap and nn.yaml are not actually operators, they are implementation details of operators, and thus don't need to participate in VariableType, JIT dispatch generation.

So, we will split these functions out from the usual Type/operator hierarchy; for now the dispatch will be done by a Type-like class called LegacyTHDispatcher.  Once this is done this probably means we can collapse Type to be backend-specific, not Type/ScalarType specific, because all the ScalarType specific code will live in the LegacyTHDispatcher.

ezyang commented on aten/src/ATen/LegacyTHDispatch.cpp:
```diff
@@ -0,0 +1,12 @@
+
+namespace at {
+
+// TODO: This could be bad juju if someone calls globalContext() in the
```
Comment out of date

gchanan replied:
is it?  This is the same comment as in globalLegacyTypeDispatch -- is it out of date there?  I'll admit I don't know what the comment is referring to.

ezyang commented on aten/src/ATen/LegacyTHDispatch.h:
```diff
@@ -0,0 +1,79 @@
+
+// LegacyTHDispatcher is the legacy mechanism for dispatching directly
+// to TH/THNN/THC/THCUNN functions in ATen, which is essentially a giant virtual
+// dispatch table for every TH function we support dynamically dispatching over.
```
I think what you want to add to the comment here is:

1. Why you need another dispatcher, as opposed to using the existing LegacyTypeDispatch (you say that it's similar to the LegacyTypeDispatch with some simplifications, but that doesn't explain the "why")
2. Under what situation we can delete this (when all the TH functions go away right)

ezyang reviewed (approved):
I'm OK with this because we've discussed it in person, but I think you need to make sure @dzhulgakov and @smessmer understand what's going on here.

gchanan committed:
Introduce LegacyTHDispatcher for dispatching to TH functions. (#14708)

Summary:
This isn't hooked up to anything yet, this is just putting the skeleton in place.
The idea here is that the functions generated via Declarations.cwrap and nn.yaml are not actually operators, they are implementation details of operators, and thus don't need to participate in VariableType, JIT dispatch generation.

So, we will split these functions out from the usual Type/operator hierarchy; for now the dispatch will be done by a Type-like class called LegacyTHDispatcher.  Once this is done this probably means we can collapse Type to be backend-specific, not Type/ScalarType specific, because all the ScalarType specific code will live in the LegacyTHDispatcher.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14708

Reviewed By: ezyang

Differential Revision: D13304654

Pulled By: ezyang

fbshipit-source-id: dac7c7fba8b58d52836f48466e986487c58e78b0
